src/Utilities/Compiler/Extensions/ITypeSymbolExtensions.cs

```diff
@@ -294,6 +294,11 @@ public static bool IsAttribute(this ITypeSymbol symbol)
         public static bool HasValueCopySemantics(this ITypeSymbol typeSymbol)
             => typeSymbol.IsValueType || typeSymbol.SpecialType == SpecialType.System_String;
 
+        public static bool HasReferenceTypeSemantics([NotNullWhen(returnValue: true)] this ITypeSymbol? typeSymbol)
+            => typeSymbol.IsReferenceTypeOrNullableValueType() ||
```

Comment:
The method name is quite unclear to me. For example, what does it mean that a nullable value type has reference type semantics?

Reply:
@Youssef1313 - do you have a recommended name suggestion, I couldn't think of a good name to indicate `is default value for this type 'null'`?

Reply:
Would `CanHoldNullValue` be good?

Reply:
Maybe we should even revise other usages of `IsReferenceTypeOrNullableValueType` and maybe move the extra cases in it?

Reply:
> Maybe we should even revise other usages of `IsReferenceTypeOrNullableValueType` and maybe move the extra cases in it?

Let me file a separate issue to audit the callsites, and make fixes with appropriate unit tests being added.